### PR TITLE
Handle type-only imports/exports for TS 3.8

### LINKED ISSUE #14: Type-only imports are converted to regular ones on sort
**Describe the bug**
TS 3.8 introduced the new feature of [Type-Only imports](https://devblogs.microsoft.com/typescript/announcing-typescript-3-8/#type-only-imports-exports) which is really for optimizing.

Unfortunately, the extension is not able to preserve it and the keyword is removed during the sort.

**To Reproduce**
```ts
import type { Client } from 'urql'
```

**Expected behavior**
Import should keep `type` keyword.

**Actual behavior**
```ts
import { Client } from 'urql'
```

**OS (please complete the following information):**
 - OS: Windows
 - Version: 10

**VS Code (please complete the following information):**
 - Version: 1.45.0

I believe the fix should be fairly trivial and I am able to work on the PR.

### PR DESCRIPTION
**Pull request by FredyC**
Initial work to fix #14, for now only a failing test.

I tried to have a look through AST how to identify `type` only import/export, but I am not seeing it in there.

https://astexplorer.net/#/gist/055e51ba88b48bc2ec5f89d7aa179240/921c3267c5b3d27edc75531a081dedbcc75274ba

### CODE REVIEW

File: src/test/suite/examples/others/type-only.origin.ts [hunk 1]
```diff
@@ -0,0 +1,4 @@
+import type A from 'A'
```
Comment (daidodo) [inline src/test/suite/examples/others/type-only.origin.ts:1]: Nit: Please use lower-case letters for paths.

**Comment by FredyC** [comment]
Got it, it's `importClause.isTypeOnly` ... I've added it to ImportNode. But I am totally lost in how to format it back to text.

**Comment by FredyC** [comment]
@daidodo Will, you be able to look into how to compose those nodes back? Sadly that code is a kinda entangled mess for me (no offense).

File: src/format/parser/ExportNode.ts [hunk 2]
```diff
@@ -13,6 +13,7 @@ import Statement, { StatementArgs } from './Statement';
 
 export default class ExportNode extends Statement {
   private readonly moduleIdentifier_?: string;
+  private readonly isTypeOnly: boolean;
```
Comment (daidodo) [inline src/format/parser/ExportNode.ts:16]: Nit: For private fields, please add `_` in the end.

File: src/format/parser/ExportNode.ts [hunk 3]
```diff
@@ -23,16 +24,18 @@ export default class ExportNode extends Statement {
       .map(getNameBinding);
     if (moduleSpecifier && moduleSpecifier.kind !== SyntaxKind.StringLiteral) return undefined;
     const moduleIdentifier = moduleSpecifier && (moduleSpecifier as StringLiteral).text;
-    return new ExportNode(moduleIdentifier, names, args);
+    return new ExportNode(moduleIdentifier, names, args, node.isTypeOnly);
```
Comment (daidodo) [inline src/format/parser/ExportNode.ts:27]: Nit: Please extract isTypeOnly in line 20.

File: src/format/parser/ImportNode.ts [hunk 4]
```diff
@@ -75,6 +75,10 @@ export default class ImportNode extends Statement {
     return this.binding_;
   }
 
+  get isTypeOnly() {
+    return (this.node_ as ImportDeclaration).importClause?.isTypeOnly ?? false;
+  }
+
```
Comment (daidodo) [inline src/format/parser/ImportNode.ts:81]: I think this field can be private, the same handling as in `ExportNode`. If so please rename it as `isTypeOnly_`.

**Comment by FredyC** [comment]
@daidodo Might be more effective if you change those nits by yourself :) As I said, I am clueless how to compose it back together.

**Comment by daidodo** [comment]
@FredyC, thanks for calling this out! I'll continue it based on your good start.

**Comment by FredyC** [comment]
> Unused: Will unused type names be removed correctly?

I don't see why not, the unused is handled in the middle of it. The `isTypeOnly` is basically a flag that tells you to include `type` to resulting declaration.

> Merge:
> * Typed default names can NOT merge with any other imports.
> * Typed name bindings can merge with other typed name bindings but not other imports.
> * Typed exports can merge with other typed exports but not regular ones.

I think this isn't a concern for this extension because if you write either of those into the user code, TS will scream at you. It should be safe to assume these rules won't be broken.

> 3\. Sort: How to sort regular and typed imports? Coming from a C++ background I lean a bit on typed imports first:

I have no problem with that.

**Comment by daidodo** [comment]
Sorry for the delay! I'm ready to merge the PR.

File: src/format/parser/ImportNode.ts [hunk 5]
```diff
@@ -253,11 +268,11 @@ export default class ImportNode extends Statement {
     const path = this.moduleIdentifier;
     const ending = quote(path);
     if (this.isScript) return `import ${ending}`;
+    const verb = 'import' + (this.isTypeOnly_ ? ' type' : '');
```
Comment (FredyC) [inline src/format/parser/ImportNode.ts:267]: Just a nit, but what about abstracting this to a single function for export/import to have it shared?

Comment (daidodo) [reply src/format/parser/ImportNode.ts:267]: Still think it's too trivial.

**Pull request merged**